icicle/backend/cpu/include/ntt_tasks_manager.h

```diff
@@ -0,0 +1,416 @@
+  {
+    if (is_full()) { return eIcicleError::OUT_OF_MEMORY; }
+    task_buffer[tail] = ntt_task_coordinates;
+    increment(tail);
+    return eIcicleError::SUCCESS;
+  }
+
+  /**
+   * @brief Retrieves a pointer to a slot for the next task coordinates.
+   *
+   * Provides access to a slot in the task buffer where new task coordinates can be assigned.
+   * @return Pointer to `NttTaskCoordinates` if a slot is available, `nullptr` otherwise.
+   */
+  template <typename S, typename E>
+  NttTaskCoordinates* NttTasksManager<S, E>::get_slot_for_next_task_coordinates()
+  {
+    if (is_full()) { return nullptr; }
+    NttTaskCoordinates* task = &task_buffer[tail];
+    increment(tail);
+    return task;
+  }
+
+  /**
+   * @brief Retrieves the next available task ready for execution.
+   *
+   * Fetches the next task from the available task buffer that is ready to be executed based on dependency
+   * resolution. If no tasks are available, returns `nullptr`.
+   *
+   * @return Pointer to `NttTaskCoordinates` of the available task, or `nullptr` if none are available.
+   */
+  template <typename S, typename E>
+  NttTaskCoordinates* NttTasksManager<S, E>::get_available_task()
+  {
+    if (head == tail) {
+      // No available tasks
+      return nullptr;
+    }
+    NttTaskCoordinates* task = &task_buffer[head];
+    increment(head);
+    return task;
+  }
+
+  /**
+   * @brief Checks if there are tasks remaining to be processed.
+   *
+   * Determines whether there are any tasks left to execute or pending dependencies.
+   *
+   * @return `true` if there are tasks to do, `false` otherwise.
+   */
+  template <typename S, typename E>
+  bool NttTasksManager<S, E>::tasks_to_do() const
+  {
+    return head != tail || nof_pending_tasks != 0;
+  }
+
+  /**
+   * @brief Checks if there are available tasks ready for execution.
+   *
+   * Determines whether there are any tasks in the buffer that are ready to be executed.
+   *
+   * @return `true` if there are available tasks, `false` otherwise.
+   */
+  template <typename S, typename E>
+  bool NttTasksManager<S, E>::available_tasks() const
+  {
+    return head != tail;
+  }
+
+  /**
+   * @brief Marks a task as completed and updates dependencies.
+   *
+   * This function should be called when a task has finished execution. It decrements the relevant
+   * dependency counters and, if dependencies are resolved, dispatches dependent tasks for execution or adds them to the
+   * task buffer.
+   *
+   * @param completed_task Pointer to the completed task.
+   * @param nof_subntts_l1 Number of sub-NTTs in the second layer of hierarchy 1.
+   * @return `true` if a dependent task was dispatched as a result of this completion, `false` otherwise.
+   */
+  template <typename S, typename E>
+  bool NttTasksManager<S, E>::handle_completed(NttTask<S, E>* completed_task, uint32_t nof_subntts_l1)
+  {
+    bool task_dispatched = false;
+    NttTaskCoordinates task_c = *completed_task->get_coordinates();
```

Comment:
why coping so many time the coordinates?
can't you use the corrdinates directly from the task?